### DISCUSSION TITLE
What is the difference between dynamic_dfs and X_df when making forecasts?

### DISCUSSION THREAD

**Discussion by davidrpugh** [Nixtla/mlforecast#230] [General]
Thanks for the awesome times series forecasting libraries! I am teaching myself your libraries by working my way through the [Kaggle Store Sales Time Series Forecasting](https://www.kaggle.com/competitions/store-sales-time-series-forecasting/code) competition.

I am trying to understand the difference between `dynamic_dfs` and `X_df` arguments to the `predict` method. What is the difference?

In the store sales competition there is a variable for oil price whose values we are given for the forecasting window. This seems like it should be included in `X_df` as the price of oil is not impacted by store sales. However, there are other variables `on promotion` and `transactions` whose values could conceivably depend on `sales`. Should these variables be included as `dynamic_dfs`?

**Comment by jmoralez** [comment]
Hey @davidrpugh, thanks for using mlforecast. They're both meant to do the same, which is providing future values of exogenous features. The `dynamic_dfs` argument is deprecated and it used to take a list of dataframes and perform merges internally, you should now use `X_df` which takes a single dataframe with id, time and the values for the exogenous features. If you're using that please upgrade to 0.10.0, since there was a bug in the `X_df` processing.

For your example I think they should all be included in `X_df`, since their values change over time. Please also make sure to set the `static_features` argument of `MLForecast.fit` accordingly, so that those features aren't considered as static. You may find [this guide](https://nixtla.github.io/mlforecast/docs/how-to-guides/exogenous_features.html) useful.

Please let us know if you have further doubts.
